R/data.table.R

```diff
@@ -2681,14 +2681,21 @@ setnames = function(x,old,new,skip_absent=FALSE) {
   invisible(x)
 }
 
-setcolorder = function(x, neworder=key(x), before=NULL, after=NULL)  # before/after #4358
+setcolorder = function(x, neworder=key(x), before=NULL, after=NULL,skip_absent=FALSE)  # before/after #4358
 {
   if (is.character(neworder) && anyDuplicated(names(x)))
     stopf("x has some duplicated column name(s): %s. Please remove or rename the duplicate(s) and try again.", brackify(unique(names(x)[duplicated(names(x))])))
   if (!is.null(before) && !is.null(after))
     stopf("Provide either before= or after= but not both")
   if (length(before)>1L || length(after)>1L)
     stopf("before=/after= accept a single column name or number, not more than one")
+  if (!isTRUEorFALSE(skip_absent))
+    stopf("skip_absent should be TRUE or FALSE")
+  if (skip_absent && is.character(neworder)){
+    neworder = intersect(neworder, colnames(x))
+  } else if (skip_absent && is.numeric(neworder)){
+    neworder = intersect(neworder, seq_along(x))
+  }
   neworder = colnamesInt(x, neworder, check_dups=FALSE)  # dups are now checked inside Csetcolorder below
```

Comment:
More general would be an argument `skip_absent=` to `colnamesInt`, then the change can be even easier:

```r
neworder = colnamesInt(x, neworder, check_dups=FALSE, skip_absent=skip_absent)
```

But I think that's better for follow-up.

Reply:
#6067